volatility3/framework/layers/qemu.py

```diff
@@ -87,15 +202,21 @@ def _get_ram_segments(self, index: int, page_size: int) -> Tuple[List[Tuple[int,
             addr = addr ^ (addr & (page_size - 1))
             index += 8
 
+            # Skip PCI-hole
+            if self._PCI_hole and addr >= self._PCI_hole[0]:
+                addr += self._PCI_hole[1]
+
             if flags & self.SEGMENT_FLAG_MEM_SIZE:
                 namelen = self._context.object(self._qemu_table_name + constants.BANG + 'unsigned char',
                                                offset = index,
                                                layer_name = self._base_layer)
                 while namelen != 0:
-                    # if base_layer.read(index + 1, namelen) == b'pc.ram':
-                    #     total_size = self._context.object(self._qemu_table_name + constants.BANG + 'unsigned long long',
-                    #                                       offset = index + 1 + namelen,
-                    #                                       layer_name = self._base_layer)
+                    name = base_layer.read(index + 1, namelen)
+                    total_size = self._context.object(self._qemu_table_name + constants.BANG + 'unsigned long long',
```

Comment:
It's a shame it needs this value, because constructing this object every time is unfortunately very slow.  I converted line 80 over to a struct statement to save a significant amount of time when parsing qemu files, so I'd avoid adding this in if possible (or use `struct`, I'd love to use volatility for all our parsing, but it's got quite a lot of overhead that isn't really necessary here).

Reply:
Wow, converting these to a `struct.unpack('>Q', ...)` indeed saves a massive amount of time! I haven't seen this change commit yet on line 80. Is that on purpose and do you still need to commit that? The same change could perhaps be applied to more places in this file, speeding it up even more.

Reply:
It's part of the patch I was rolling for #587 so hasn't been committed yet.  I was trying not to put it in too many places, because otherwise people question what benefit they get from volatility parsing the data.  I need to look at ways of making object creation more efficient, but for now I only wanted to do as few changes as possible.  I think there's two object creations that could be replace with `struct` in the big loop.  The others will get call a couple times and make next to no difference.

Reply:
If you want to get it in sooner, I'm happy to review just that as a PR and we can probably merge that pretty quickly.  Since we're in code-freeze though it's unlikely to make it into the next release tarball.  I'm hoping the one after this will only be 6 months away rather than a year, but just so you know the way we expect things to go...

Reply:
> It's part of the patch I was rolling for #587 so hasn't been committed yet. I was trying not to put it in too many places, because otherwise people question what benefit they get from volatility parsing the data. I need to look at ways of making object creation more efficient, but for now I only wanted to do as few changes as possible. I think there's two object creations that could be replace with `struct` in the big loop. The others will get call a couple times and make next to no difference.

I've added the improvement to line 80 (now line 202-203). What was the other location you intended to change?

Reply:
Hmmm, probably in `_load_segments` I'd check `section_byte` since it's used for every loop.  The easiest way to check is to run `python -m cProfile -o out.prof vol.py <some vol command involving qemu>`.  That'll create an out.prof, then you can use something like [Snakeviz](https://jiffyclub.github.io/snakeviz/) to see where the time's being spent.  That'll help you look around for what's using the most time, it boils down to calls to `object` in the `_get_ram_segments` method, but it could be any of the calls there.  After converting the addr call, you can see how much of an effect it had, and then see whether it's worth doing to the two `namelen` object calls as well...

![image](https://user-images.githubusercontent.com/1477845/160508472-8e288764-e087-456d-b2e0-3270545c0337.png)

Reply:
Thanks for the explanation @ikelos! I tried to use `struct.unpack` at different other locations, but the performance impact is quite low (<0.5 second faster when applied to multiple locations). Compared to the performance gain that is achieved by parsing the address using `struct.unpack`, I think we should leave the other locations as they are.